vignettes/cmdstanr-internals.Rmd

```diff
@@ -0,0 +1,293 @@
+```
+
+### JSON file
+
+If you already have your data in a JSON file you can just pass that file
+directly to CmdStanR instead of using a list of R objects. For example, we could
+pass in the JSON file we created above using `write_stan_json()`:
+
+```{r data-json, eval=FALSE}
+fit <- mod$sample(data = json_file)
+```
+
+
+### R dump file
+
+Finally, it is also possible to use the R dump file format. This is *not*
+recommended because CmdStan can process JSON faster than R dump, but CmdStanR
+allows it because CmdStan will accept files created by `rstan::stan_rdump()`:
+
+```{r data-rdump, eval=FALSE}
+rdump_file <- tempfile(fileext = ".data.R")
+rstan::stan_rdump(names(data_list), file = rdump_file, envir = list2env(data_list))
+cat(readLines(rdump_file), sep = "\n")
+fit <- mod$sample(data = rdump_file)
+```
+
+
+## Writing CmdStan output to CSV
+
+### Default temporary files
+
+```{r sample-tempdir, results = "hide"}
+data_list <- list(N = 10, y = c(0,1,0,0,0,0,0,0,0,1))
+fit <- mod$sample(data = data_list)
+```
+
+When fitting a model, the default behavior is to write the output from CmdStan 
+to CSV files in a temporary directory:
+
+```{r output_files}
+fit$output_files()
+```
+
+These files will be lost if you end your R session or if you remove the 
+`fit` object and force (or wait for) garbage collection. 
+
+
+```{r gc}
+files <- fit$output_files()
+file.exists(files)
+
+rm(fit)
+gc()
+
+file.exists(files)
+```
+
+### Non-temporary files
+
+To save these files to a non-temporary location there are two options. You 
+can either specify the `output_dir` argument to `mod$sample()` or use 
+`fit$save_output_files()` after fitting the model: 
+
+
+```{r save_output_files, eval=FALSE}
+# see ?save_output_files for info on optional arguments
+fit$save_output_files(dir = "path/to/directory")
+```
+
+```{r output_dir, eval = FALSE}
+fit <- mod$sample(
+  data = data_list, 
+  output_dir = "path/to/directory"
+)
+```
+
+
+## Reading CmdStan output into R
+
+### Lazy CSV reading 
+
+With the exception of some diagnostic information, the CSV files are not read
+into R until their contents are requested by calling a method that requires them
+(e.g., `fit$draws()`, `fit$summary()`, etc.). If we examine the structure of the
+`fit` object, notice how the `Private` slot `draws_` is `NULL`, indicating that
+the CSV files haven't yet been read into R:
+
+```{r refit, include=FALSE}
+fit <- mod$sample(data = data_list)
+```
+```{r csv-not-read}
+str(fit)
+```
+
+After we call a method that requires the draws then if we reexamine the 
+structure of the object we will see that the `draws_` slot in `Private` 
+is no longer empty: 
+
+```{r for-csv-reading}
+draws <- fit$draws() # force CSVs to be read into R
+str(fit)
+```
+
+For models with many parameters, if only some parameters are requested (e.g., by
+specifying the `variables` argument to `fit$draws()`) then CmdStanR will only
+read in the requested variables (unless they have already been read in).
```

Comment:
Optional: We could also close https://github.com/stan-dev/cmdstanr/issues/174 by mentioning that in order to store all samples one should call

`fit$draws()`

and then save the fit object.

Reply:
Good idea. I'll add that.

Reply:
Actually, I think it would be necessary to also call `fit$sampler_diagnostics()` to make sure those get saved too. I'll add that also. 

But should we just add a `save` method to the fitted model objects that forces reading everything in before saving? It could look like this:

```r
save = function(file, ...) {
  self$draws()
  self$sampler_diagnostics()
  saveRDS(self, file, ...)
}
```

Reply:
That would probably be the best yes. Great idea!

Reply:
Ok I added a method `fit$save_object()`. In doing that I fixed a big bug that meant that we were actually reading in all draws again even if they had already been read in! See comment below. So I'm glad we decided to add the save method or I wouldn't have noticed!